lib/ImportFromC/HeaderToModel.cpp

```diff
@@ -619,50 +619,71 @@ bool DeclVisitor::VisitFunctionDecl(const clang::FunctionDecl *FD) {
+      if (not ParamType) {
+        Errors.emplace_back("revng: Unsupported type for a raw function "
+                            "argument.\n");
+        return false;
+      }
+
       std::optional Register = parseStringAnnotation(*ParamDecl,
                                                      RegAnnotation,
                                                      Errors);
+      std::optional Stack = parseStringAnnotation(*ParamDecl,
+                                                  StackAnnotation,
+                                                  Errors);
       if (not Register.has_value()) {
-        std::optional Stack = parseStringAnnotation(*ParamDecl,
-                                                    StackAnnotation,
-                                                    Errors);
-        if (I != N - 1) {
-          Errors.emplace_back("revng: Only the very last RFT argument is "
-                              "allowed to represent stack.\n");
-          return false;
-        }
-
-        if (Stack.has_value()) {
-          // TODO: Handle stack location.
-          //
-          // TODO: Don't forget to assert that it is always a pointer.
-          Errors.emplace_back("revng: We don't support stack parameters in "
-                              "RFTs for now.\n");
-          return false;
-        } else {
+        if (not Stack.has_value()) {
           Errors.emplace_back("revng: Arguments must have either a register or "
                               "a stack annotation.\n");
           return false;
+        } else {
+          if (I != N - 1) {
+            Errors.emplace_back("revng: Only the very last RFT argument is "
+                                "allowed to represent stack (which also means "
+                                "there can only be one! If you need to "
+                                "represent multiple values, use a struct).\n");
+            return false;
+          }
+
+          if (not ParamType->isStruct()) {
+            Errors.emplace_back("revng: RFT stack argument must be a "
+                                "struct.\n");
+            return false;
+          }
+
+          if (not TheRawFunctionType.StackArgumentsType().isEmpty()) {
+            Errors.emplace_back("revng: Multiple stack arguments? How did this "
+                                "even happen?.\n");
+            return false;
+          }
+
+          TheRawFunctionType.StackArgumentsType() = std::move(ParamType);
+          if (ParamDecl->getName() != "stack") {
+            Errors.emplace_back("revng: WARNING: stack argument name ('"
+                                + ParamDecl->getName().str()
+                                + "') was ignored.\n");
+          }
+        }
+      } else {
+        if (Stack.has_value()) {
+          Errors.emplace_back("revng: WARNING: _STACK annotation ignored "
+                              "because _REG takes precedence.\n");
```

Comment:
My two cents is that better error reporting is something the codebase severely lacks. I've been postponing this conversation because, in @aleclearmind's words, we have bigger fish to fry.

But we could definitely benefit from a more standardized error-warning reporting solution adopted through the whole pipeline.

It feel so bad, for example, to just write something to a logger when raw->cabi conversion fails - it's not like anyone but me is going to even look at that logger :D. If we had some centralized place to gather all warnings that user might want to see (bonus points for being able to group them by tiers / severity / where they are from, etc). Maybe even a log-like file appended to by the pipeline run that UI exposes or something. This needs some thought put into it for sure, but we need something. And I'd prefer for that something to be more handy that "vector of `llvm::Error`s".

-----------

As for the current state of `HeaderToModel`, we specifically use the brittle-testing @pfez mentioned. Which, in this case, doesn't feel _that_ brittle:
1. Run the analysis on a given model with a given c-input.
2. If it succeeds, run `revng model compare` on the resulting model.
3. If it fails, compare its output against `error.txt` file included with the test (with stripped license header).

Reply:
Yeah in this specific situation I think that "pinning" error messages is acceptable until we don't do a proper redesign.

Speaking of which @Cvelth, given that you've added new failure modes, do all of them have associated tests with expected failure?

Reply:
There's definitely one for every important failure, but if you prefer a more complete coverage, it shouldn't be hard to add more. Especially now that they are directory segregated.

Reply:
I think it's pretty clear that error reporting to user in edit types needs a lot of love.
But this is out of scope as of now.

For testing, clang does various stuff but they also hardcode error messages: https://github.com/llvm/llvm-project/blob/main/clang/docs/analyzer/developer-docs/DebugChecks.rst
It's not ideal, but I think it's fine.

I'm adding a task about improving error reporting (in edit types, but also in verify) for which we'll schedule the priority. But let's focus on the goal of this PR for now: adding support for `_START` and `_SIZE`.

Reply:
> if you prefer a more complete coverage, it shouldn't be hard to add more.

Yes please. I'd like at least one test that tests expected failure for each error message. Thanks!

> But let's focus on the goal of this PR for now

Sure, I stated it upfront. I just wanted to get the discussion started and agree on the best course of action now.
Thanks @aleclearmind

Reply:
I made as many new tests as felt reasonable (see the caveats in the mattermost message for specifics). Please take a look at the very last commit.

Reply:
Closing this as you did indeed take a look (thanks).